exonum-java-binding/testkit/src/main/java/com.exonum.binding.testkit/TestKit.java

```diff
@@ -0,0 +1,197 @@
+    } catch (NoSuchMethodException e) {
+      throw new IllegalArgumentException("No no-arg constructor", e);
+    } catch (InstantiationException | InvocationTargetException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private native long nativeCreateTestKit(UserServiceAdapter[] services, boolean auditor,
+      short withValidatorCount, TimeProvider timeProvider);
+
+  private native long nativeCreateSnapshot(long nativeHandle);
+
+  private native Block nativeCreateBlock(long nativeHandle);
+
+  private native Block nativeCreateBlockWithTransactions(long nativeHandle, byte[][] transactions);
+
+  private native EmulatedNode nativeGetEmulatedNode(long nativeHandle);
+
+  /**
+   * Creates a new builder for the TestKit.
+   *
+   * @param nodeType type of the main TestKit node - either validator or auditor. Note that
+   * {@link Service#afterCommit(BlockCommittedEvent)} logic will only be called on the main TestKit node of this
+   * type
+   */
+  public static Builder builder(EmulatedNodeType nodeType) {
+    checkNotNull(nodeType);
+    return new Builder(nodeType);
+  }
+
+  /**
+   * Builder for the TestKit.
+   */
+  public static final class Builder {
+
+    private EmulatedNodeType nodeType;
+    private short validatorCount;
+    private List<Class<? extends ServiceModule>> services = new ArrayList<>();
+    private TimeProvider timeProvider;
+
+    private Builder(EmulatedNodeType nodeType) {
+      this.nodeType = nodeType;
+    }
+
+    /**
+     * Sets number of additional validator nodes in the TestKit network. Note that
+     * {@link Service#afterCommit(BlockCommittedEvent)} logic will only be called on the main TestKit node.
+     */
+    public Builder withValidators(short validatorCount) {
+      this.validatorCount = validatorCount;
+      return this;
+    }
+
+    /**
+     * Adds a service with which the TestKit would be instantiated. Several services can be added.
+     */
+    public Builder withService(Class<? extends ServiceModule> serviceModule) {
+      services.add(serviceModule);
+      return this;
+    }
+
+    /**
+     * Adds a list of services with which the TestKit would be instantiated. Several services can
+     * be added.
+     */
+    public Builder withServices(List<Class<? extends ServiceModule>> serviceModules) {
+      services.addAll(serviceModules);
+      return this;
+    }
+
+    /**
+     * If called, will create a TestKit with time service with given TimeProvider.
+     */
+    public Builder withTimeService(TimeProvider timeProvider) {
+      this.timeProvider = timeProvider;
+      return this;
+    }
+
+    /**
+     * Creates the TestKit instance.
+     */
+    public TestKit build() {
+      checkMaxServiceNumber(services);
+      return new TestKit(services, nodeType, validatorCount, timeProvider);
+    }
+
+    private void checkMaxServiceNumber(List<Class<? extends ServiceModule>> serviceModules) {
```

Comment:
Is there anything else that must be validated? E.g., non-empty list of services?

Reply:
I added such a check, even though it's possible to create a blockchain without any user services and if users want to run such a network (only with time service, for example), maybe we should let them, even if it's useless? But probably it's better to fail in such case to indicate that a user forgot to add a service